types/react/test/hooks.tsx

```diff
@@ -0,0 +1,78 @@
+import * as React from "react";
+
+interface PersonProps {
+    name: string;
+    age: number;
+}
+export function Person(props: PersonProps) {
+    return <div>
+        hello! I'm {props.name} and I'm {props.age} years old!
+    </div>;
+}
+
+export interface FancyButtonProps {
+    onClick: () => void;
+    children?: React.ReactNode;
+}
+export interface FancyButton {
+    fancyClick(): void;
+}
+export const FancyButton = React.forwardRef((props: FancyButtonProps, ref: React.Ref<FancyButton>) => {
+    const buttonRef = React.useRef<HTMLButtonElement | null>(null);
+    const [count, setCount] = React.useState(0);
+
+    React.useImperativeMethods(ref, () => ({
+        fancyClick() {
+            buttonRef.current!; // $ExpectType HTMLButtonElement
+        },
+        getClickCount() {
+            return count;
+        }
+    }));
+
+    return <button onClick={() => { setCount(count + 1); props.onClick(); }}>
+        {props.children}
+    </button>;
+});
+
+interface AppState {
+    name: string;
+    age: number;
+}
+
+type AppActions =
+    | { type: "getOlder" }
+    | { type: "resetAge" }
```

Comment:
A convenient thing of this vs redux is that these don't have to be objects at all. Could just be `enum AppActions { getOlder, resetAge }` 😄